tests/upgrade/pre/scenario/e2e/runtime/trust-root-km-upgrade.go

```diff
@@ -0,0 +1,97 @@
+package runtime
+
+import (
+	"context"
+	"fmt"
+
+	governance "github.com/oasisprotocol/oasis-core/go/governance/api"
+	"github.com/oasisprotocol/oasis-core/go/oasis-test-runner/env"
+	"github.com/oasisprotocol/oasis-core/go/oasis-test-runner/oasis"
+	"github.com/oasisprotocol/oasis-core/go/oasis-test-runner/scenario"
+	"github.com/oasisprotocol/oasis-core/go/oasis-test-runner/scenario/e2e/runtime"
+	staking "github.com/oasisprotocol/oasis-core/go/staking/api"
+
+	"github.com/oasisprotocol/oasis-core/test-upgrade/scenario/e2e"
+)
+
+var (
+	// TrustRootKeymanagerUpgrade is the keymanager upgrade pre-upgrade scenario.
+	TrustRootKeymanagerUpgrade scenario.Scenario = newTrustRootKeymanagerUpgradeImpl()
+
+	trustRootKeymanagerUpgradeTestClientScenario = runtime.NewTestClientScenario([]interface{}{
+		runtime.RuntimeInsertKeyValueTx{Key: "key1", Value: "value1", Response: "", Encrypted: false},
+		runtime.RuntimeInsertKeyValueTx{Key: "key2", Value: "value2", Response: "", Encrypted: true},
+		runtime.RuntimeGetKeyValueTx{Key: "key1", Response: "value1", Encrypted: false},
+		runtime.RuntimeGetKeyValueTx{Key: "key1", Response: "", Encrypted: true},
+		runtime.RuntimeGetKeyValueTx{Key: "key2", Response: "", Encrypted: false},
+		runtime.RuntimeGetKeyValueTx{Key: "key2", Response: "value2", Encrypted: true},
+	})
+)
+
+type trustRootKeymanagerUpgradeImpl struct {
+	runtime.TrustRootImpl
+}
+
+func newTrustRootKeymanagerUpgradeImpl() scenario.Scenario {
+	return &trustRootKeymanagerUpgradeImpl{
+		TrustRootImpl: *runtime.NewTrustRootImpl(
+			"keymanager-upgrade",
+			runtime.NewTestClient().WithScenario(trustRootKeymanagerUpgradeTestClientScenario),
+		),
+	}
+}
+
+func (sc *trustRootKeymanagerUpgradeImpl) Clone() scenario.Scenario {
+	return &trustRootKeymanagerUpgradeImpl{
+		TrustRootImpl: *sc.TrustRootImpl.Clone().(*runtime.TrustRootImpl),
+	}
+}
+
+func (sc *trustRootKeymanagerUpgradeImpl) Fixture() (*oasis.NetworkFixture, error) {
+	f, err := sc.TrustRootImpl.Fixture()
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure entities can vote.
+	f.Network.FundEntities = true
+	f.Network.StakingGenesis = &staking.Genesis{}
+
+	// Speed up voting.
+	f.Network.GovernanceParameters = &governance.ConsensusParameters{
+		StakeThreshold:            67,
+		UpgradeCancelMinEpochDiff: 3,
+		UpgradeMinEpochDiff:       3,
+		VotingPeriod:              2,
+	}
+
+	// Keep runtime bundles for post-upgrade scenario.
+	for i := range f.Runtimes {
+		f.Runtimes[i].KeepBundles = true
+	}
+
+	return f, nil
+}
+
+func (sc *trustRootKeymanagerUpgradeImpl) Run(childEnv *env.Env) error {
+	ctx := context.Background()
+
+	// Build a simple key/value runtime with embedded trust root, start the network, register
+	// runtimes and run the test client.
```

Comment:
I assume this does not also build the key manager with trust root? We should do that as well to make sure the key manager also survives (should be the same as runtime, but in production both have trust roots configured).

Reply:
That is correct. In fact, our key manager tests have three issues that need to be addressed:
1. We have not been testing the key manager runtime with an embedded trust root, as mentioned above.
2. We are testing insecure operations that do not require the key manager at all. For example, the trust root scenario uses the unencrypted `SimpleKeyValueScenario`.
3. We need to ensure that the compute nodes clear the cache whenever changes are made to the key manager. This is optional because currently, a faulty key manager can cause tests to fail with a consensus failure if at least one of the nodes has an empty cache.

These problems will be addressed in another pull request. The focus is on setting up the environment for upgrade tests.

Reply:
Agreed, I've created tasks for these items.